pkg/reconciler/contour/resources/kingress.go

```diff
@@ -40,6 +43,45 @@ func MakeEndpointProbeIngress(ctx context.Context, ing *v1alpha1.Ingress) *v1alp
 	}
 
 	sns := ServiceNames(ctx, ing)
+
+	// Reverse engineer our previous state from the prior generation's HTTP Proxy resources.
+	for _, proxy := range previousState {
+		// Establish the visibility based on the class annotation.
+		var vis v1alpha1.IngressVisibility
+		for v, class := range config.FromContext(ctx).Contour.VisibilityClasses {
+			if class == proxy.Annotations["projectcontour.io/ingress.class"] {
+				vis = v
+			}
+		}
+		if vis == "" {
+			continue
+		}
+
+		for _, route := range proxy.Spec.Routes {
+			for _, svc := range route.Services {
+				si, ok := sns[svc.Name]
+				if !ok {
+					si = ServiceInfo{
+						Port:         intstr.FromInt(svc.Port),
+						Visibilities: make([]v1alpha1.IngressVisibility, 0, 1),
+					}
+				}
+				has := false
+				for _, v := range si.Visibilities {
+					if v == vis {
+						has = true
+						break
+					}
+				}
+				if !has {
+					si.Visibilities = append(si.Visibilities, vis)
+				}
+				sns[svc.Name] = si
+			}
+		}
+	}
+
+	// Give the services a deterministic ordering.
 	order := make(sets.String, len(sns))
 	for key := range sns {
 		order.Insert(key)
```

Comment:
Can you have dups there? If not, putting into slice and `sort.Strings` is probably a better op

Reply:
There can be duplicates.